File: typescipt/existing/src/language/expressions/functions/lookupRowFunction.ts






namespace Lexy.Compiler.Language.Expressions.Functions;

internal class LookupRowFunction : ExpressionFunction, IHasNodeDependencies
{
   private const string FunctionHelp = " Arguments: LOOKUPROW(Table, lookUpValue, Table.SearchValueColumn)";

   public const string Name = "LOOKUPROW";

   private const int Arguments = 3;
   private const int ArgumentTable = 0;
   private const int ArgumentLookupValue = 1;
   private const int ArgumentSearchValueColumn = 2;

   public string Table { get; }

   public Expression ValueExpression { get; }

   public MemberAccessLiteral SearchValueColumn { get; }

   public VariableType SearchValueColumnType { get; private set; }
   public VariableType RowType { get; private set; }

   private LookupRowFunction(string tableType, Expression valueExpression,
     MemberAccessLiteral searchValueColumn, SourceReference tableNameArgumentReference)
     : base(tableNameArgumentReference)
   {
     Table = tableType ?? throw new ArgumentNullException(nameof(tableType));
     ValueExpression = valueExpression ?? throw new ArgumentNullException(nameof(valueExpression));
     SearchValueColumn = searchValueColumn ?? throw new ArgumentNullException(nameof(searchValueColumn));
   }

   public IEnumerable<IRootNode> GetDependencies(RootNodeList rootNodeList)
   {
     var table = rootNodeList.GetTable(Table);
     if (table ! null) yield return table;
   }

   public static ParseExpressionFunctionsResult Parse(string name, SourceReference functionCallReference,
     IReadOnlyList<Expression> arguments)
   {
     if (arguments.Count ! Arguments)
       return ParseExpressionFunctionsResult.Failed($"Invalid number of arguments. {FunctionHelp}");

     if (!(arguments[ArgumentTable] is IdentifierExpression tableNameExpression))
       return ParseExpressionFunctionsResult.Failed(
         $"Invalid argument {ArgumentTable}. Should be valid table name. {FunctionHelp}");

     if (!(arguments[ArgumentSearchValueColumn] is MemberAccessExpression searchValueColumnHeader))
       return ParseExpressionFunctionsResult.Failed(
         $"Invalid argument {ArgumentSearchValueColumn}. Should be search column. {FunctionHelp}");

     var tableName = tableNameExpression.Identifier;
     var valueExpression = arguments[ArgumentLookupValue];
     var searchValueColumn = searchValueColumnHeader.MemberAccessLiteral;

     var lookupFunction =
       new LookupRowFunction(tableName, valueExpression, searchValueColumn, functionCallReference);
     return ParseExpressionFunctionsResult.Success(lookupFunction);
   }

   public override IEnumerable<INode> GetChildren()
   {
     yield return ValueExpression;
   }

   protected override void Validate(IValidationContext context)
   {
     ValidateColumn(context, SearchValueColumn, ArgumentSearchValueColumn);

     var tableType = context.RootNodes.GetTable(Table);
     if (tableType = null)
     {
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentTable}. Table name '{Table}' not found. {FunctionHelp}");
       return;
     }

     var searchColumnHeader = tableType.Header.Get(SearchValueColumn);
     if (searchColumnHeader = null)
     {
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentSearchValueColumn}. Column name '{SearchValueColumn}' not found in table '{Table}'. {FunctionHelp}");
       return;
     }

     var conditionValueType = ValueExpression.DeriveType(context);
     SearchValueColumnType = searchColumnHeader.Type.CreateVariableType(context);

     if (conditionValueType = null | !conditionValueType.Equals(SearchValueColumnType))
       context.Logger.Fail(Reference,
         $"Invalid argument {ArgumentSearchValueColumn}. Column type '{SearchValueColumn}': '{SearchValueColumnType}' doesn't match condition type '{conditionValueType}'. {FunctionHelp}");

     RowType = tableType?.GetRowType(context);
   }


   private void ValidateColumn(IValidationContext context, MemberAccessLiteral column, int index)
   {
     if (column.Parent ! Table)
       context.Logger.Fail(Reference,
         $"Invalid argument {index}. Result column table '{column.Parent}' should be table name '{Table}'");

     if (column.Parts.Length ! 2)
       context.Logger.Fail(Reference,
         $"Invalid argument {index}. Result column table '{column.Parent}' should be table name '{Table}'");
   }

   public override VariableType DeriveReturnType(IValidationContext context)
   {
     var tableType = context.RootNodes.GetTable(Table);
     return tableType?.GetRowType(context);
   }
}

File: src/Lexy.Compiler/Language/VariableTypes/Functions/LookupRowFunction.cs
using System.Collections.Generic;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Language.Functions;
using Lexy.Compiler.Language.Tables;
using Lexy.Compiler.Parser;

namespace Lexy.Compiler.Language.VariableTypes.Functions;

internal class LookUpRowFunction : TableFunctionReference
{
    private const string FunctionHelpValue =
        "Arguments: " +
           "TableName.LookUpRow(lookUpValue) " +
        "or TableName.LookUpRow(lookUpValue, Table.SearchColumn)" +
        "or TableName.LookUpRow(discriminator, lookUpValue)" +
        "or TableName.LookUpRow(discriminator, lookUpValue, Table.DiscriminatorColumn, Table.SearchColumn)";

    private record OverloadArguments(
        int? Discriminator,
        int LookUpValue,
        int? DiscriminatorColumnArgument,
        int? DefaultDiscriminatorColumn,
        int? SearchColumnArgument,
        int DefaultSearchColumn) : IOverloadArguments;

    public const string Name = "LookUpRow";

    protected override string FunctionHelp => FunctionHelpValue;

    internal LookUpRowFunction(Table table): base(table)
    {
    }

    public override VariableType GetResultsType(IReadOnlyList<Expression> arguments) => Table?.GetRowType();

    public override ValidateInstanceFunctionArgumentsResult ValidateArguments(IValidationContext context, IReadOnlyList<Expression> arguments,
        SourceReference reference)
    {
         if (!ValidateTable(context, reference)) return ValidateInstanceFunctionArgumentsResult.Failed();

        var overloadArguments = GetArgumentColumns(context, arguments, reference);
        if (overloadArguments == null) return null;

        var searchColumnHeader = GetColumn(context, arguments, overloadArguments.SearchColumnArgument, overloadArguments.DefaultSearchColumn, reference);

        if (searchColumnHeader == null) return ValidateInstanceFunctionArgumentsResult.Failed();

        ValidateColumnValueType(context, arguments, overloadArguments.LookUpValue, "Search", searchColumnHeader, reference);

        var discriminatorColumnHeader = ValidatorDiscriminator(context, arguments, reference, overloadArguments);

        var result = new LookUpRowFunctionCall(
            Table.Name.Value,
            arguments[overloadArguments.LookUpValue],
            overloadArguments.Discriminator.HasValue ? arguments[overloadArguments.Discriminator.Value] : null,
            searchColumnHeader.Name,
            discriminatorColumnHeader?.Name);

        return ValidateInstanceFunctionArgumentsResult.Success(result);
    }

    private OverloadArguments GetArgumentColumns(IValidationContext context, IReadOnlyList<Expression> arguments, SourceReference reference)
    {
        switch (arguments.Count)
        {
            case 1:
                //"table.LookUpRow(lookUpValue)"
                return new OverloadArguments(null, 0, null, null, null, 0);

            case 2:
                //"table.LookUpRow(lookUpValue, Table.SearchColumn)"
                if (arguments[1] is MemberAccessExpression)
                {
                    return new OverloadArguments(null, 0, null, null, 1, 0);
                }
                //"table.LookUpRow(discriminator, lookUpValue)"
                return new OverloadArguments(0, 1, null, 0, null, 1);

            case 4:
                //"table.LookUpRow(discriminator, lookUpValue, Table.DiscriminatorColumn, Table.SearchColumn)";
                return new OverloadArguments(0, 1, 2, 0, 3, 1);

            default:
                context?.Logger.Fail(reference, $"Invalid number of arguments. {FunctionHelpValue}");
                return null;
        }
    }
}